iceoryx_utils/include/iceoryx_utils/internal/cxx/storable_function.hpp

```diff
@@ -0,0 +1,165 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef IOX_UTILS_STORABLE_FUNCTION_HPP
+#define IOX_UTILS_STORABLE_FUNCTION_HPP
+
+#include "iceoryx_utils/cxx/function_ref.hpp"
+#include "iceoryx_utils/cxx/type_traits.hpp"
+#include "iceoryx_utils/internal/cxx/static_storage.hpp"
+
+#include <iostream>
+#include <type_traits>
+#include <utility>
+
+namespace iox
+{
+namespace cxx
+{
+template <typename ReturnType, typename... Args>
+using signature = ReturnType(Args...);
+
+template <typename StorageType, typename T>
+class storable_function;
+
+/// @brief A storable alternative of std::function which not necessarily uses dynamic memory.
+/// @param StorageType The type of internal storage to store the actual data.
+///                    Needs to provide allocate and deallocate functions.
+///                    See static_storage.hpp for a static memory version.
+/// @param ReturnType  The return type of the stored callable.
+/// @param Args        The arguments of the stored callable.
+template <typename StorageType, typename ReturnType, typename... Args>
+class storable_function<StorageType, signature<ReturnType, Args...>>
+{
+  public:
+    using signature_t = signature<ReturnType, Args...>;
+
+    storable_function() noexcept = default;
+
+    /// @brief construct from functor (including lambda)
+    template <typename Functor,
+              typename = typename std::enable_if<std::is_class<Functor>::value
+                                                     && is_invocable_r<ReturnType, Functor, Args...>::value,
+                                                 void>::type>
+    storable_function(const Functor& functor) noexcept;
+
+    /// @brief construct from function pointer (including static functions)
+    storable_function(ReturnType (*function)(Args...)) noexcept;
+
+    /// @brief construct from object reference and member function
+    /// only a pointer to the object is stored for the call
+    template <typename T, typename = typename std::enable_if<std::is_class<T>::value, void>::type>
+    storable_function(T& object, ReturnType (T::*method)(Args...)) noexcept;
+
+    /// @brief construct from object reference and const member function
+    /// only a pointer to the object is stored for the call
+    template <typename T, typename = typename std::enable_if<std::is_class<T>::value, void>::type>
+    storable_function(const T& object, ReturnType (T::*method)(Args...) const) noexcept;
+
+    storable_function(const storable_function& other) noexcept;
+
+    storable_function(storable_function&& other) noexcept;
+
+    storable_function& operator=(const storable_function& rhs) noexcept;
+
+    storable_function& operator=(storable_function&& rhs) noexcept;
+
+    ~storable_function() noexcept;
+
+    /// @brief invoke the stored function
+    /// @note  invoking the function if there is no stored function (i.e. operator bool returns false)
```

Comment:
Completely against changing the signature to returning ``expected<ReturnType`` for efficiency and semantic reasons (it is not what the user wanted, otherwise he would have said so).

Now, we can check the condition internally, yes. I can live with this but I consider the STL approach (which I am following here) superior (same goes for ``function_ref`` but I did not implement this).

I will put an assert like construct here.

Reply:
Uses Expects now, leading to defined termination.